### PR TITLE
Implement locate-dominating-file and dir-locals

### PR DESCRIPTION
**Pull request by CeleritasCelery**
This is based on [jsadusk/tramp-hlo](https://github.com/jsadusk/tramp-hlo). That repo provided huge speedups to a few functions that currently don't have tramp methods. I implemented these in tramp-rpc and made them a single remote call.

# Benchmarks

| Test | RPC-with-hlo (new) | RPC-no-hlo (old) | sshx | vs SSH | vs no-hlo |
|---|---|---|---|---|---|
| `locate-dominating-file` | **69.70 ms** | 232.41 ms | 240.51 ms | 3.45x | 3.33x |
| `dir-locals-find-file` | **70.81 ms** | 308.39 ms | 729.26 ms | **10.30x** | 4.36x |
| `dir-locals--all-files` | **69.52 ms** | 223.39 ms | 426.48 ms | 6.13x | 3.21x |

### CODE REVIEW

**Review by ArthurHeymans — COMMENTED** [review]
Thanks for this. I tried to get it working with LLM and these suggestions came out. WDYT?

File: lisp/tramp-rpc-advice.el [hunk 1]
```diff
@@ -441,6 +482,12 @@ so that .dir-locals.el files are detected and loaded normally."
   (with-eval-after-load 'magit-process
     (advice-add 'magit-start-process :around
                 #'tramp-rpc--magit-start-process-advice))
+  (advice-add 'locate-dominating-file :around
+              #'tramp-rpc--locate-dominating-file-advice)
+  (advice-add 'dir-locals-find-file :around
+              #'tramp-rpc--dir-locals-find-file-advice)
```
Comment (ArthurHeymans) [inline lisp/tramp-rpc-advice.el]:
```suggestion
;; NOTE: dir-locals-find-file advice is intentionally NOT installed.
;; The stock `dir-locals-find-file' calls `locate-dominating-file' with
;; `#'dir-locals--all-files' as a function predicate.  The
;; `locate-dominating-file' advice correctly falls through to the original
;; implementation for function predicates, and `dir-locals--all-files' is
;; advised separately below to use the RPC fast-path.  This avoids a
;; subtle interaction where the custom handler returns nil.
```

File: lisp/tramp-rpc-advice.el [hunk 2]
```diff
@@ -416,6 +419,44 @@ so that .dir-locals.el files are detected and loaded normally."
...
+;; operation registration APIs.
+(defun tramp-rpc--locate-dominating-file-advice (orig-fun file name)
+  "Use RPC fast-path for remote `locate-dominating-file'."
+  (if (and (stringp file)
+           (tramp-rpc-file-name-p
+            (if (file-name-absolute-p file)
+                file
+              (file-name-concat default-directory file))))
+      (tramp-rpc-handle-locate-dominating-file file name)
+    (funcall orig-fun file name)))
```
Comment (ArthurHeymans) [inline lisp/tramp-rpc-advice.el]: Only intercepts calls where NAME is a string (file name to search for).
Function predicates fall through to the original implementation to avoid
infinite recursion via `tramp-run-real-handler'.
File-name-quoted paths (with /: prefix) also fall through since the
handler does not support them."
```suggestion
  (let ((expanded (if (and (stringp file) (file-name-absolute-p file))
                      file
                    (file-name-concat default-directory file))))
    (if (and (not (functionp name))
             (stringp file)
             (not (file-name-quoted-p expanded))
             (tramp-rpc-file-name-p expanded))
        (tramp-rpc-handle-locate-dominating-file file name)
      (funcall orig-fun file name))))
```

**Comment by CeleritasCelery** [comment]
@ArthurHeymans finally got this passing. I am glad you have so many CI checks to keep bugs from slipping through. You probably want to squash this before merging 😁

**Comment by ArthurHeymans** [comment]
> @ArthurHeymans finally got this passing. I am glad you have so many CI checks to keep bugs from slipping through. You probably want to squash this before merging 😁

Tramp has a lot of tests and I got a lot of help from the tramp maintainer to get more of them passing :-)

It looks like this needs a manual rebase again.

**Comment by ArthurHeymans** [comment]
tramp introduced a new way to add advice: https://github.com/ArthurHeymans/emacs-tramp-rpc/pull/148 would you mind also using it?

**Comment by CeleritasCelery** [comment]
initially I did, but the tramp version that adds `tramp-add-external-operation` is not included with Emacs until version 31 which has not been release yet. Users would have to be using the upstream version of tramp for that feature to work. see also [this issue](https://github.com/jsadusk/tramp-hlo/issues/5). I assumed since this package only requires Emacs 30.1 that we would not want to use the upstream feature. If we do change it we should make sure to tell the users they need to be using tramp 2.8 or later.

**Comment by ArthurHeymans** [comment]
> initially I did, but the tramp version that adds `tramp-add-external-operation` is not included with Emacs until version 31 which has not been release yet. Users would have to be using the upstream version of tramp for that feature to work. see also [this issue](https://github.com/jsadusk/tramp-hlo/issues/5). I assumed since this package only requires Emacs 30.1 that we would not want to use the upstream feature. If we do change it we should make sure to tell the users they need to be using tramp 2.8 or later.

Wouldn't package managers handle the tramp dependency automatically?

**Comment by CeleritasCelery** [comment]
> Wouldn't package managers handle the tramp dependency automatically?

I have only used [straight](https://github.com/radian-software/straight.el) and [elpaca](https://github.com/progfolio/elpaca), and they will not download external versions of builtin packages even if a newer version is required by the package manifest. I am not sure about others. The user would need to explicitly declare tramp as an external package in their config.

**Comment by ArthurHeymans** [comment]
> > Wouldn't package managers handle the tramp dependency automatically?
> 
> I have only used [straight](https://github.com/radian-software/straight.el) and [elpaca](https://github.com/progfolio/elpaca), and they will not download external versions of builtin packages even if a newer version is required by the package manifest. I am not sure about others. The user would need to explicitly declare tramp as an external package in their config.

Right those won't do it, but I think the default package manager would.
I added the following https://github.com/ArthurHeymans/emacs-tramp-rpc/pull/154 documents the requirement and adds a check in the code.

**Comment by CeleritasCelery** [comment]
@ArthurHeymans  I switched to the external operation method. Let me know if everything else looks good.

File: server/src/handlers/commands.rs [hunk 3]
```diff
@@ -203,3 +202,236 @@ pub async fn ancestors_scan(params: Value) -> HandlerResult {
...
+}
+
+fn mtime_seconds(path: &Path) -> Option<i64> {
+    let modified = std::fs::metadata(path).ok()?.modified().ok()?;
+    let duration = modified.duration_since(UNIX_EPOCH).ok()?;
+    Some(duration.as_secs() as i64)
+}
+
+/// Return readable regular files from a directory for a list of names.
+pub async fn highlevel_test_files_in_dir(params: &Value) -> HandlerResult {
```
Comment (ArthurHeymans) [inline server/src/handlers/commands.rs]: All three new handlers take `params: &Value` (reference), but every other handler in the codebase takes `params: Value` (owned). This forces:
- `from_value(params.clone())` inside each handler (unnecessary allocation)
- `&params` in the dispatch table in `mod.rs` instead of the uniform `params`

The `from_value` wrapper at `protocol.rs:522` takes `Value` (owned), so the clone exists only to work around the `&Value` signature.

Suggestion: change all three to `fn ...(params: Value) -> HandlerResult` and drop the `.clone()` calls, matching the pattern of `run_parallel`, `ancestors_scan`, and every other handler.

File: server/src/handlers/commands.rs [hunk 4]
```diff
@@ -203,3 +202,236 @@ pub async fn ancestors_scan(params: Value) -> HandlerResult {
...
+
+fn as_search_dir(path: &Path) -> Option<PathBuf> {
+    if path.is_dir() {
+        Some(path.to_path_buf())
+    } else {
+        path.parent().map(|p| p.to_path_buf())
+    }
+}
+
+fn find_dominating_dir(start_dir: &Path, names: &[String]) -> Option<(PathBuf, Vec<String>)> {
```
Comment (ArthurHeymans) [inline server/src/handlers/commands.rs]: `find_dominating_dir` loops all the way to the filesystem root with no depth limit. The existing `ancestors_scan` handler (line 134) has a configurable `max_depth` (default 10) for the same ancestor-walking pattern. Consider adding a similar bound here for consistency and defensive coding.

File: server/src/handlers/commands.rs [hunk 5]
```diff
@@ -203,3 +202,236 @@ pub async fn ancestors_scan(params: Value) -> HandlerResult {
...
+        let file_value = resolved_file.to_string_lossy().to_string().into_value();
+
+        let Some(existing_start) = find_existing_start(&file_path) else {
+            return Ok(msgpack_map! {
+                "file" => file_value,
+                "locals" => Value::Nil,
+                "cache" => Value::Nil
+            });
+        };
+        let Some(start_dir) = as_search_dir(&canonical_or_original(existing_start)) else {
```
Comment (ArthurHeymans) [inline server/src/handlers/commands.rs]: Inconsistent symlink handling between the two ancestor-walking handlers:

- `highlevel_locate_dominating_file_multi` (line 306) intentionally does **not** canonicalize: `as_search_dir(existing_start)`, with a comment about preserving lexical symlink paths.
- Here, the same start point **is** canonicalized: `as_search_dir(&canonical_or_original(existing_start))`.

This means for a file opened through a symlink, `locate-dominating-file` returns the symlink-relative directory, but `dir-locals-find-file` returns the canonical (resolved) directory. Since `dir-locals-directory-cache` is keyed by directory path, the cache entries produced here may not match paths from `locate-dominating-file`, which could cause stale/missed cache lookups.

Should this also preserve the lexical path, or should `locate_dominating_file_multi` also canonicalize? Either way, they should be consistent.

File: lisp/tramp-rpc.el [hunk 6]
```diff
@@ -1758,6 +1761,152 @@ ID-FORMAT specifies whether to use numeric or string IDs."
...
+                 (tramp-rpc--quote-localname
+                  quoted-localname
+                  (tramp-rpc--decode-string path))))
+              result))))
+
+(defun tramp-rpc-handle-locate-dominating-file (file name)
+  "Like `locate-dominating-file' for TRAMP-RPC files.
+For string/list NAME, uses a high-level RPC call.  Predicate NAME falls back
+to the built-in implementation."
+  ;; This intentionally ignores `locate-dominating-stop-dir-regexp'.
```
Comment (ArthurHeymans) [inline lisp/tramp-rpc.el]: This is a behavioral difference vs stock `locate-dominating-file`, which checks `locate-dominating-stop-dir-regexp` at each ancestor step. Users who set that variable (e.g., to avoid traversing past `/home` or into network mount parents) will get different results with the RPC method.

Since the server-side `find_dominating_dir` doesn't know about Elisp variables, a practical fix could be:
1. Check the stop regexp on the Elisp side against the returned directory (post-RPC validation), or
2. Pass the regexp to the server and apply it in the Rust loop, or
3. At minimum, document this as a known limitation in the docstring.

Option 1 is simplest -- if the dominating directory is above the stop point, return nil.

File: lisp/tramp-rpc.el [hunk 7]
```diff
@@ -1758,6 +1761,152 @@ ID-FORMAT specifies whether to use numeric or string IDs."
...
+                      (file-name-as-directory
+                       (concat file-connection
+                               (tramp-rpc--decode-string
+                                (alist-get 'dir cache-dir-update))))))
+         (dir-elt (when cache-dir
+                    (seq-find (lambda (elt) (string= (car elt) cache-dir))
+                              dir-locals-directory-cache))))
+    (if (and dir-elt
+             (or (null locals-dir)
+                 (<= (length locals-dir) (length (car dir-elt)))))
```
Comment (ArthurHeymans) [inline lisp/tramp-rpc.el]: Using `(length locals-dir)` vs `(length (car dir-elt))` as a proxy for "which directory is deeper" works when both paths share a common prefix (since both are ancestors of `file`), but string length doesn't reliably correspond to directory depth. A path like `/a/very-long-dirname/` is longer than `/a/b/c/d/` despite being shallower.

Consider comparing component counts instead, e.g.:
```elisp
(<= (length (file-name-split locals-dir))
    (length (file-name-split (car dir-elt))))
```
Or using `file-in-directory-p` to check containment directly.

**Review by ArthurHeymans — COMMENTED** [review]
## Review: Implement locate-dominating-file and dir-locals

The core feature is solid -- the benchmarks are compelling (3-10x), the parity tests comparing optimized handlers against `tramp-run-real-handler` are excellent, the symlink preservation test catches a real edge case, and the path quoting/unquoting logic matches existing conventions well.

See inline comments for specific issues. Summary of findings:

### Correctness
1. **`params: &Value` breaks codebase convention** -- every other handler takes owned `Value`, forcing unnecessary `.clone()` in each new handler (see inline)
2. **Symlink canonicalization inconsistency** between `locate_dominating_file_multi` (preserves lexical path) and `dir_locals_find_file_cache_update` (canonicalizes) -- could cause cache key mismatches (see inline)
3. **`locate-dominating-stop-dir-regexp` silently ignored** -- behavioral difference vs stock Emacs (see inline)
4. **No depth limit on `find_dominating_dir`** -- existing `ancestors_scan` has `max_depth`, this walks to root unconditionally (see inline)

### Design
5. **Path length as proxy for directory depth** in `tramp-rpc-handle-dir-locals-find-file` is fragile (see inline)
6. **Removed `tramp-rpc-mock-test-dir-locals-advice-installed` test** -- the `hack-dir-local-variables` advice is still active in `tramp-rpc-advice.el`, but the test verifying its install/remove cycle was deleted and not replaced. The new tests cover the handler functions but not the advice mechanism itself.

**Comment by CeleritasCelery** [comment]
@ArthurHeymans I addressed your feedback. Please review

**Comment by ArthurHeymans** [comment]
Awesome, thx a lot!
